vnext/ReactUWP/Views/KeyboardEventHandler.cpp

```diff
@@ -0,0 +1,317 @@
+  : KeyboardEventBaseHandler(keyDown, keyUp)
+{
+}
+
+void PreviewKeyboardEventHandler::hook(XamlView xamlView)
+{
+  auto uiElement = xamlView.as<winrt::UIElement>();
+  if (m_keyDownCallback)
+    m_previewKeyDownRevoker = uiElement.PreviewKeyDown(winrt::auto_revoke, m_keyDownCallback);
+
+  if (m_keyUpCallback)
+    m_previewKeyUpRevoker = uiElement.PreviewKeyUp(winrt::auto_revoke, m_keyUpCallback);
+}
+
+void PreviewKeyboardEventHandler::unhook()
+{
+  m_previewKeyDownRevoker.revoke();
+  m_previewKeyUpRevoker.revoke();
+}
+
+KeyboardEventHandler::KeyboardEventHandler(KeyboardEventCallback keyDown, KeyboardEventCallback keyUp)
+  :KeyboardEventBaseHandler(keyDown, keyUp)
+{
+}
+
+void KeyboardEventHandler::hook(XamlView xamlView)
+{
+  auto uiElement = xamlView.as<winrt::UIElement>();
+  if (m_keyDownCallback)
+    m_KeyDownRevoker = uiElement.KeyDown(winrt::auto_revoke, m_keyDownCallback);
+
+  if (m_keyUpCallback)
+    m_KeyUpRevoker = uiElement.KeyUp(winrt::auto_revoke, m_keyUpCallback);
+}
+
+void KeyboardEventHandler::unhook()
+{
+  m_KeyDownRevoker.revoke();
+  m_KeyUpRevoker.revoke();
+}
+
+PreviewKeyboardEventHandlerOnRoot::PreviewKeyboardEventHandlerOnRoot(const std::weak_ptr<IReactInstance>& reactInstance)
+  :PreviewKeyboardEventHandler(
+    std::bind(&PreviewKeyboardEventHandlerOnRoot::OnPreKeyDown, this, _1, _2),
+    std::bind(&PreviewKeyboardEventHandlerOnRoot::OnPreKeyUp, this, _1, _2)),
+  m_wkReactInstance(reactInstance)
+{
+}
+
+void PreviewKeyboardEventHandlerOnRoot::OnPreKeyDown(winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const&args)
+{
+  DispatchEventToJs("topKeyDown", args);
+}
+
+void PreviewKeyboardEventHandlerOnRoot::OnPreKeyUp(winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const&args)
+{
+  DispatchEventToJs("topKeyUp", args);
+}
+
+inline bool IsModifiedKeyPressed(winrt::CoreWindow const& coreWindow, winrt::VirtualKey vk)
+{
+  return (coreWindow.GetKeyState(vk) & winrt::CoreVirtualKeyStates::Down) == winrt::CoreVirtualKeyStates::Down;
+}
+
+inline bool IsModifiedKeyLocked(winrt::CoreWindow const& coreWindow, winrt::VirtualKey vk)
+{
+  return (coreWindow.GetKeyState(vk) & winrt::CoreVirtualKeyStates::Locked) == winrt::CoreVirtualKeyStates::Locked;
+}
+
+template<typename T> void UpdateModifiedKeyStatusTo(T& ev)
+{
+  auto const& coreWindow = winrt::CoreWindow::GetForCurrentThread();
+  ev.altKey = IsModifiedKeyPressed(coreWindow, winrt::VirtualKey::Menu);
+  ev.shiftKey = IsModifiedKeyPressed(coreWindow, winrt::VirtualKey::Shift);
+  ev.metaKey = IsModifiedKeyPressed(coreWindow, winrt::VirtualKey::LeftWindows)
+    || IsModifiedKeyPressed(coreWindow, winrt::VirtualKey::RightWindows);
+  ev.ctrlKey = IsModifiedKeyPressed(coreWindow, winrt::VirtualKey::Control);
+  ev.capLocked = IsModifiedKeyLocked(coreWindow, winrt::VirtualKey::CapitalLock);
+};
+
+void PreviewKeyboardEventHandlerOnRoot::DispatchEventToJs(string eventName, winrt::KeyRoutedEventArgs args)
```

Comment:
>string [](start = 58, length = 6)

Pass by reference

Reply:
agree

Reply:
eventName is still being passed by value.

---
In reply to: [295534425](https://github.com/microsoft/react-native-windows/pull/2623#discussion_r295534425) [](ancestors = 295534425)